grafana/grafana
# Feature: Trusted Types support

To try this PR, enable the feature flag `trustedTypes`.

[Trusted Types](https://www.w3.org/TR/trusted-types/) is security mechanism that locks down DOM API's that often provides sinks that are used when exploiting DOM XSS. By using trusted types you will force the string to be handled (preferably sanitized) before the string can be used by the API. Trusted types provides a robust defense against DOM XSS. Trusted types is currently only supported in Chrome.

This PR shows a minimal working PoC that uses `createHTML`, `createScript` and `createScriptURL`. Currently each one of them uses a different method; sanitizing via a library (DOMPurify), inline sanitizing and do nothing.  

Grafana uses many third-party libraries that we can not control, but we do have some first party code that we can change to either return a trusted type object or refactor the code to not use unsafe DOM API's. These will be documented in this PR.

- [x] Choose if the code should be refactored, sanitized through the "default" policy or create a new policy
- [x] Find 1st party code that uses dangerous API's 

## Added things
1. Enable CSP for e2e tests
   -  Add meta tag with CSP content in order to make it compatible with Cypress
   - Export replace function from Go middleware
2. Add trusted types default policy
   - Initiliazed before jQuery and Angular is loaded
   - Limited sanitization in order to make Angular work
3. Add sanitization functions 
   - Returns TrustedHTML, which is trusted types compatible
   - Extend DOMParser type to support TrustedHTML 


## How to use

### Enable Trusted Types without implementation (to see errors)
1. Enable `content_security_policy_report_only = true`
7. Add `require-trusted-types-for 'script';` to `content_security_policy_report_only_template` 
8. Set `disable_sanitize_html = true` in order to see errors

### Test implementation
1. Set feature flag `trustedTypes = true`
2. See no errors and removed HTML
